src/Workspaces/Core/Portable/Workspace/ProjectSystem/FileWatchedReferenceFactory.cs

```diff
@@ -4,127 +4,98 @@
     /// </remarks>
-    private readonly ConditionalWeakTable<PortableExecutableReference, string> _previousDisposalLocations = new();
+    private readonly ConditionalWeakTable<TReference, string> _previousDisposalLocations = new();
 
     /// <summary>
-    /// <see cref="CancellationTokenSource"/>s for in-flight refreshing of metadata references. When we see a file change, we wait a bit before trying to actually
-    /// update the workspace. We need cancellation tokens for those so we can cancel them either when a flurry of events come in (so we only do the delay after the last
-    /// modification), or when we know the project is going away entirely.
+    /// <see cref="CancellationTokenSource"/>s for in-flight refreshing of metadata references. When we see a file
+    /// change, we wait a bit before trying to actually update the workspace. We need cancellation tokens for those so
+    /// we can cancel them either when a flurry of events come in (so we only do the delay after the last modification),
+    /// or when we know the project is going away entirely.
     /// </summary>
-    private readonly Dictionary<string, CancellationTokenSource> _metadataReferenceRefreshCancellationTokenSources = [];
+    private readonly Dictionary<string, CancellationTokenSource> _referenceRefreshCancellationTokenSources = [];
 
-    public FileWatchedPortableExecutableReferenceFactory(
-        IFileChangeWatcher fileChangeWatcher)
+    /// <summary>
+    /// Notification when a reference changes on disk.  The string value is the full file path of the changed reference.
+    /// </summary>
+    public event EventHandler<string>? ReferenceChanged;
+
+    public FileWatchedReferenceFactory(IFileChangeWatcher fileChangeWatcher, Func<ImmutableArray<WatchedDirectory>> getAdditionalWatchedDirectories)
     {
         _fileReferenceChangeContext = new Lazy<IFileChangeContext>(() =>
         {
-            var referenceDirectories = new HashSet<string>();
-
-            // On each platform, there is a place that reference assemblies for the framework are installed. These are rarely going to be changed
-            // but are the most common places that we're going to create file watches. Rather than either creating a huge number of file watchers
-            // for every single file, or eventually realizing we should just watch these directories, we just create the single directory watchers now.
-            // We'll collect this from two places: constructing it from known environment variables, and also for the defaults where those environment
-            // variables would usually point, as a fallback.
-
-            if (Environment.GetEnvironmentVariable("DOTNET_ROOT") is string dotnetRoot && !string.IsNullOrEmpty(dotnetRoot))
-            {
-                referenceDirectories.Add(Path.Combine(dotnetRoot, "packs"));
-            }
-
-            if (PlatformInformation.IsWindows)
```

Comment:
this moved to logic only run for the PortableExecutionReference case, not the AnalyzerFileReference case.  I'm not 100% sure if that's what we want, or if we should always be watching tehse special folders for both cases.  @jasonmalinowski  @dibarbet ?

Reply:
where does an AnalyzerFileReference come from? (e.g. what is the user scenario)?

Reply:
I assume we shouldn't be creating multiple file watchers for the same directories and file kinds though (I don't think we should have multiple file watchers and then file change events for the same directory and types)?

Reply:
it generally comes from a user having an `<AnalyzerReference>` in their project file.  :)

Reply:
hm - I would think then this probably only applies to PortableExecutableReference and probably isn't needed for AnalyzerFileReference.  Not 100% sure on that though

Reply:
yeah, i'll talk to @jasonmalinowski  to make sure.

Reply:
Thanks!

Reply:
I would expect we _do_ have this same logic for both, although it may have less of a benefit in this case. The reason for these general directory watches is we expect a lot of references to come from them, and since they don't really change often it saves us a lot of overhead watching each file separately. If the SDK ends up with a dozen or so analyzers (if it doesn't already?) multiply that across a bunch of projects and you have a nice win.

Hopefully it's easy enough to refactor?